l3backend/l3backend-color.dtx

```diff
@@ -1248,6 +1253,401 @@
 %</package>
 %    \end{macrocode}
 %
+% \subsection{Color in \LuaTeX{}}
+%
+%    \begin{macrocode}
+%<*lua>
+%    \end{macrocode}
+%
+% In \LuaTeX{}, we can use an approach based on attributes to avoid having to
+% insert whatsits. That means having the appropriate supporting Lua code. The
+% underlying idea comes from Heiko Oberdiek's \pkg{luacolor} package.
+%
+% \begin{macro}{ltx.color}
+% \begin{macro}{ltxcolor}
+%   The basic table should be set up in \pkg{l3luatex}, but we want a separate
+%   sub-table for color.
+%    \begin{macrocode}
+ltx = ltx or {color = {}}
+ltx.color = ltx.color or {}
+local ltxcolor = ltx.color
+%    \end{macrocode}
+% \end{macro}
+% \end{macro}
+%
+% Load basic support and print some information.
+%    \begin{macrocode}
+require("ltluatex")
+local luatexbase = luatexbase
+luatexbase.provides_module({name = "l3backend-luatex", date = "2021-06-03"})
+%    \end{macrocode}
+%
+% \begin{macro}{attr}
+%   An attribute to track color. There is no need to access this from \TeX{} so
+%   in contrast to \pkg{luacolor} we just allocate directly in Lua.
+%    \begin{macrocode}
+local color_attr = luatexbase.new_attribute("color")
+%    \end{macrocode}
+% \end{macro}
+%
+% \begin{macro}{map}
+%   A map between color strings and their color_index, plus the tracking number.
+%    \begin{macrocode}
+local map = {n = 0}
+%    \end{macrocode}
+% \end{macro}
+%
+% \begin{macro}{color_index}
+%   Get the color_index of a color or create a new color_index if necessary.
+%    \begin{macrocode}
+local function color_index(color)
+  local n = map[color]
+  if not n then
+    n = map.n + 1
+    map[n] = color
+    map[color] = n
+  end
+  return n
+end
```

Comment:
Not important for the first version, but I think `color_attr` (read-only) and `color_index` should be available for external code. One of the less obvious but still important advantages of the attribute approach is that this can be used with nodes which are generated programmatically.

Reply:
I was aiming to keep interfaces to a minimum at least at this stage. I assume you are thinking about using directly from Lua, in which case wouldn't we still keep the `attr` and the `index` private and provide `ltx.color.set` or something?

Reply:
> I assume you are thinking about using directly from Lua, in which case wouldn't we still keep the `attr` and the `index` private and provide `ltx.color.set` or something?

I think that having direct access is also useful since it allows e.g. to be more precise when inheriting attributes (when creating a node between two other nodes, it can be useful to only inherit their color attribute if both are colored the same / if at least one is colored etc. to avoid unnecessary color changes, especially when you know that it wouldn't have an effect on your node anyway). Not sure about it though.



> I was aiming to keep interfaces to a minimum at least at this stage.

Fully agreed.

Reply:
I 

> > I assume you are thinking about using directly from Lua, in which case wouldn't we still keep the `attr` and the `index` private and provide `ltx.color.set` or something?
> 
> I think that having direct access is also useful since it allows e.g. to be more precise when inheriting attributes (when creating a node between two other nodes, it can be useful to only inherit their color attribute if both are colored the same / if at least one is colored etc. to avoid unnecessary color changes, especially when you know that it wouldn't have an effect on your node anyway). Not sure about it though.

 I guess what I mean is that they are just numbers, and those should really always be internal representations of abstract concepts. TeX is already bad enough with `\catcode` ....